core/transformer/src/ECReferenceTypesCache.ts

```diff
@@ -0,0 +1,169 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module iModels
+ */
+
+import { DbResult, Logger, TupleKeyedMap } from "@itwin/core-bentley";
+import { ConcreteEntityTypes, IModelError, RelTypeInfo } from "@itwin/core-common";
+import { ECClass, Mixin, RelationshipClass, RelationshipConstraint, Schema, SchemaKey, SchemaLoader, StrengthDirection } from "@itwin/ecschema-metadata";
+import * as assert from "assert";
+import { IModelDb } from "@itwin/core-backend";
+
+const logger = Logger.makeCategorizedLogger("ECClassNavPropReferenceCache");
+
+/** The context for transforming a *source* Element to a *target* Element and remapping internal identifiers to the target iModel.
+ * @internal
+ */
+export class SchemaNotInCacheErr extends Error {
+  public constructor() { super("Schema was not in cache, initialize that schema"); }
+}
+
+/**
+ * A cache of the entity types referenced by navprops in ecchemas, as well as the source and target entity types of
+ * The transformer needs the referenced type to determine how to resolve references.
+ * @internal
+ */
+export class ECReferenceTypesCache {
+  /** singleton because using multiple of these is mostly reinitializing duplicated info expensively */
+  public static globalCache = new ECReferenceTypesCache();
```

Comment:
Is singleton necessary vs creating a cache for a specific iModel?

Reply:
it isn't cheap so I opted to use a singleton.

Reply:
Unbounded caches aren't cheap either, and the mixing of data from different versions of the same schema seems a little error-prone. Does a typical user of the transformer create multiple IModelCloneContexts?

Reply:
Is this cache gauranteed to be consistent between iModels?

Reply:
Sort of, but no. Each transformation makes one `IModelCloneContext`. The transformer tests run several transformations, and iirc testing got noticeably longer when they had to do this for each transformation. Also @ColinKerr it's not possible to change the types of EC properties or the root/abstract constraints of relationship classes? So this shouldn't be error-prone.

Reply:
Don't care about tests.

Reply:
I will talk to @ColinKerr about having a cache with 1 entry per property in the ec schemas.

Reply:
I have made it per-transformer, so it is bounded by the amount of properties in the source iModel.
Also dynamic schemas could cause bugs so I removed the global cache